### ISSUE TITLE
BUG: Program plugin does not search correctly in `Start Menu` folder

### ISSUE THREAD

**Issue by TheBestPessimist** [Flow-Launcher/Flow.Launcher#1859]
### Checks

- [X] I have checked that this issue has not already been reported.

### Problem Description

From what i can see, it searches in `Start Menu/Programs` but not `Start Menu` directly.

Desired search path: 
```
C:\Users\TheBestPessimist\AppData\Roaming\Microsoft\Windows\Start Menu
```

For example, it doesnt find this shortcut: 
```
"C:\Users\TheBestPessimist\AppData\Roaming\Microsoft\Windows\Start Menu\Zoner Photo Studio X.lnk"
```

### To Reproduce

1. Search for a program

### Screenshots

![image](https://user-images.githubusercontent.com/4482210/215312366-c81bfa0f-4cd4-44d9-8bf8-962dd755dfef.png)

### Flow Launcher Version

1.12

### Windows Build Number

win 11

### Error Log

<details>

```shell

Replace this line with the important log contents.

```
</details>

**Comment by TheBestPessimist** [comment]
Bug created for myself with SS (screenshot) so that i wont forget about it. 
Could be an easy fix?

**Comment by TheBestPessimist** [comment]
I'm trying to understand the logic of [`Flow.Launcher.Plugin.Program.Main.InitAsync`](https://github.com/Flow-Launcher/Flow.Launcher/blob/dev/Plugins/Flow.Launcher.Plugin.Program/Main.cs#L73-L106)

We're doing this async, so in theory we can be as slow as we need to (as slow as we can be).

1. Read all the cached data from disk
2. make a boolean knowing if the cache is empty or not
3. (without checking the boolean)
    - load Win32 programs async (even though were in async task anyway)
    - load UWP programs asyc (even though were in async task anyway)
4. if the cache is empty 
        - await loading
    else
        - leave the 2 runs above do whatever they want, nobody cares about them anyway
5. watch win32 and uwp programs for updates


My problem with the logic is
- why do we run those 2 loads async but we ignore whatever happens to them in case we have cache? why not cancel them?
- why do we persist the cache? how long does it take to reindex on every FL starup the win32/uwp apps? 
    - clicking reindex in setting on my laptop takes 0.5 sec maybe (however i am a dev and have an i7-9750h so i might be in minority)
    - at point 5, when we watch, that thing basically does a reindex anyway


This startup logic has another problem: say FL crashes, i install a new software, then restart FL (or restart laptop which restarts FL): at this point FL cannot know of the new software, simply because there is no way for it to be announced that something changed on my system in the past. So me as a user am forced to click reindex.

**Comment by TheBestPessimist** [comment]
I figured out what the problems to my original bug are:
1. `Search In Program Description` is `false` by default
2. Search in `Start Menu` actually searches in `Start Menu/Programs` (fixed in #1864)
3. Registry search is done before start menu search and therefore there is no `LocalisedName`, because `LocalisedName` is computed only for shortcuts (see https://github.com/TheBestPessimist/Flow.Launcher/blob/2e0923d25bdef9bc1d943ffceed53918dcdc5726/Plugins/Flow.Launcher.Plugin.Program/Programs/Win32.cs#L335-L335)

I have a few points of discussion:

1. Would it make sense to set default  `Search In Program Description`  to `true`? Would it hurt anyone?
3. This is an interesting one:
    - name of the exe: `ZPS.exe`
    - name of start shortcut: `Zoner Photo Studio X.lnk`
    - description of exe: `Zoner Photo Studio X`
    
With #1864, start menu search is working as expected and my shortcut is found. 

2 flags influence this:
- search in description: yes/no
- registry search: yes/no

| start search flag | registry flag | description flag | FL search         | Full name in FL             | Explanation                                                                             |
|-------------------|---------------|------------------|-------------------|-----------------------------|-----------------------------------------------------------------------------------------|
| ✅                 | ❌             | ❌                | `zoner`           | `Zoner Photo Studio X`      | LocalisedName taken from lnk                                                            |
| ✅                 | ❌             | ✅                | `ZPS` or `zoner`  | `ZPS: Zoner Photo Studio X` | LocalisedName taken from lnk                                                            |
| ✅                 | ✅             | ❌                | `ZPS`             | `ZPS`                       | Name of exe from registry overrides name of lnk from start                              |
| ✅                 | ✅             | ✅                | `ZPS`  or `zoner` | `ZPS: Zoner Photo Studio X` | Name of exe from registry overrides name of lnk from start, and concat with description |

**Comment by TheBestPessimist** [comment]
![image](https://user-images.githubusercontent.com/4482210/215492058-8bd475cc-b453-4de9-a4db-0c03a69cac23.png)

**Comment by taooceros** [comment]
Thank you so much!

>     * why do we run those 2 loads async but we ignore whatever happens to them in case we have cache? why not cancel them?

Actually it is performing the index for every single new startup, but just await them (prevent flow from showing up) when cache is not available.

> why do we persist the cache? how long does it take to reindex on every FL starup the win32/uwp apps?

>    clicking reindex in setting on my laptop takes 0.5 sec maybe (however i am a dev and have an i7-9750h so i might be in minority)
>    at point 5, when we watch, that thing basically does a reindex anyway

You are really the minority. It depends on how fast the disk is running. For my computer (i7 8765u) it takes around 3-4 seconds (acceptable but not so fast).


> From what i can see, it searches in Start Menu/Programs but not Start Menu directly.

I am curious what kind of program really stored in StartMenu? You seems to be able to index zoner photo studio?

**Comment by TheBestPessimist** [comment]
> Actually it is performing the index for every single new startup, but just await them (prevent flow from showing up) when cache is not available.


Yes, that is what i said as well. My question is why not cancel if cache exists?

pseudo code i suggest:

```kotlin
if !cache
    taskLoad1.cancelExecution // do not waste resources
    taskLoad2.cancelExecution // do not waste resources
else
    await taskLoad1 taskLoad2
```


> I am curious what kind of program really stored in StartMenu

[`Zoner Photo Studio X`](https://www.zoner.com) (see picture in OP) `¯\_(ツ)_/¯`


> You seems to be able to index zoner photo studio?

yes, with my merge request :).

---

Do you have opinions about my other suggestions and comments?

**Comment by taooceros** [comment]
> > Actually it is performing the index for every single new startup, but just await them (prevent flow from showing up) when cache is not available.
> 
> Yes, that is what i said as well. My question is why not cancel if cache exists?
> 
> pseudo code i suggest:
> 
> ```kotlin
> if cache
>     taskLoad1.cancelExecution // do not waste resources
>     taskLoad2.cancelExecution // do not waste resources
> else
>     await taskLoad1 taskLoad2
> ```
> 
> > I am curious what kind of program really stored in StartMenu
> 
> [`Zoner Photo Studio X`](https://www.zoner.com) (see picture in OP for path) `¯\_(ツ)_/¯`
> 
> > You seems to be able to index zoner photo studio?
> 
> yes, with my merge request :).
> 
> Do you have opinions about my other suggestions and comments?

To accommodate he situation you have mentioned :)

> This startup logic has another problem: say FL crashes, i install a new software, then restart FL (or restart laptop which restarts FL): at this point FL cannot know of the new software, simply because there is no way for it to be announced that something changed on my system in the past. So me as a user am forced to click reindex.
